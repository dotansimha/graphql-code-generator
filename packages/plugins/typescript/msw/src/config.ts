import { RawClientSideBasePluginConfig, ClientSideBasePluginConfig } from '@graphql-codegen/visitor-plugin-common';

/**
 * @description This plugin generates `msw` (https://github.com/mswjs/msw) mock handlers with TypeScript typings.
 */
export interface MSWConfig {
  /**
   * @description GraphQL endpoint to use when working with multiple backends.
   *
   * @see https://mswjs.io/docs/api/graphql/link
   *
   * @exampleMarkdown
   * ```yml
   *   config:
   *     link:
   *       name: stripe
   *       endpoint: https://api.stripe.com/graphql
   * ```
   */
  link?: {
    endpoint: string;
    name: string;
  };
}

export interface MSWRawPluginConfig extends RawClientSideBasePluginConfig, MSWConfig {}
export interface MSWPluginConfig extends ClientSideBasePluginConfig, MSWConfig {}
