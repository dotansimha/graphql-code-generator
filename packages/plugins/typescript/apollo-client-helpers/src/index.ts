import { Types, PluginValidateFn, PluginFunction } from '@graphql-codegen/plugin-helpers';
import { GraphQLSchema, isInterfaceType, isObjectType } from 'graphql';
import { extname } from 'path';
import { ApolloClientHelpersConfig } from './config';

export const plugin: PluginFunction<ApolloClientHelpersConfig> = (
  schema: GraphQLSchema,
  documents: Types.DocumentFile[],
  config: ApolloClientHelpersConfig
) => {
  const results: Types.ComplexPluginOutput[] = [];

  results.push(generateTypePoliciesSignature(schema, config));

  return {
    prepend: results.reduce((prev, r) => [...prev, ...(r.prepend || [])], []),
    append: results.reduce((prev, r) => [...prev, ...(r.append || [])], []),
    content: results.map(r => r.content).join('\n'),
  };
};

function generateTypePoliciesSignature(
  schema: GraphQLSchema,
  config: ApolloClientHelpersConfig
): Types.ComplexPluginOutput {
  const typeMap = schema.getTypeMap();
  const perTypePolicies: string[] = [];
  const typedTypePolicies = Object.keys(typeMap).reduce((prev, typeName) => {
    const type = typeMap[typeName];

    if (!typeName.startsWith('__') && (isObjectType(type) || isInterfaceType(type))) {
      const fieldsNames = Object.keys(type.getFields()).filter(f => !f.startsWith('__'));
      const keySpecifierVarName = `${typeName}KeySpecifier`;
      const fieldPolicyVarName = `${typeName}FieldPolicy`;

      perTypePolicies.push(
        `export type ${keySpecifierVarName} = (${fieldsNames
          .map(f => `'${f}'`)
          .join(' | ')} | ${keySpecifierVarName})[];`
      );

      perTypePolicies.push(`export type ${fieldPolicyVarName} = {
${fieldsNames.map(fieldName => `\t${fieldName}?: FieldPolicy<any> | FieldReadFunction<any>`).join(',\n')}
};`);

      return {
        ...prev,
        [typeName]: `Omit<TypePolicy, "fields" | "keyFields"> & {
\t\tkeyFields${
          config.requireKeyFields ? '' : '?'
        }: false | ${keySpecifierVarName} | (() => undefined | ${keySpecifierVarName}),
\t\tfields?: ${fieldPolicyVarName},
\t}`,
      };
    }

    return prev;
  }, {} as Record<string, string>);

  const rootTypes = [
    schema.getQueryType()?.name,
    schema.getMutationType()?.name,
    schema.getSubscriptionType()?.name,
  ].filter(Boolean);

  const rootContent = `export type TypedTypePolicies = TypePolicies & {${Object.keys(typedTypePolicies)
    .map(typeName => {
      const nonOptional = config.requirePoliciesForAllTypes && !rootTypes.includes(typeName);

      return `\n\t${typeName}${nonOptional ? '' : '?'}: ${typedTypePolicies[typeName]}`;
    })
    .join(',')}\n};`;

  return {
    prepend: [
      `import ${
        config.useTypeImports ? 'type ' : ''
      }{ FieldPolicy, FieldReadFunction, TypePolicies, TypePolicy } from '@apollo/client/cache';`,
    ],
    content: [...perTypePolicies, rootContent].join('\n'),
  };
}

export const validate: PluginValidateFn<ApolloClientHelpersConfig> = async (
  schema: GraphQLSchema,
  documents: Types.DocumentFile[],
  config,
  outputFile: string
) => {
  if (extname(outputFile) !== '.ts' && extname(outputFile) !== '.tsx') {
    throw new Error(`Plugin "apollo-client-helpers" requires extension to be ".ts" or ".tsx"!`);
  }
};
