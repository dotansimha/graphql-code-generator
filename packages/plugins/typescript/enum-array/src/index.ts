import { GraphQLSchema, GraphQLEnumType } from 'graphql';
import { PluginFunction, Types } from '@graphql-codegen/plugin-helpers';
import { EnumArrayPluginConfig } from './config';

function getEnumTypeMap(schema: GraphQLSchema): GraphQLEnumType[] {
  const typeMap = schema.getTypeMap();
  const result: GraphQLEnumType[] = [];
  for (const key in typeMap) {
    if (typeMap[key].astNode?.kind === 'EnumTypeDefinition') {
      result.push(typeMap[key] as GraphQLEnumType);
    }
  }
  return result;
}

function buildArrayDefinition(e: GraphQLEnumType): string {
  const upperName = e.name
    .replace(/[A-Z]/g, letter => `_${letter}`)
    .slice(1)
    .toUpperCase();
  const values = e
    .getValues()
    .map(v => `'${v.value}'`)
    .join(', ');
  return `export const ${upperName}: ${e.name}[] = [${values}];`;
}

function buildImportStatement(enums: GraphQLEnumType[], importFrom: string): string[] {
  const names: string[] = Object.values(enums).map(e => e.name);
  return [`import { ${names.join(', ')} } from "${importFrom}";`];
}

export const plugin: PluginFunction<EnumArrayPluginConfig> = async (
  schema: GraphQLSchema,
  _documents: Types.DocumentFile[],
  config: EnumArrayPluginConfig
): Promise<Types.PluginOutput> => {
  const importFrom: EnumArrayPluginConfig['importFrom'] = config.importFrom;
  const enums = getEnumTypeMap(schema);
  const content = enums.map(buildArrayDefinition).join('\n');

  const result: Types.PluginOutput = { content };
  if (importFrom) {
    result['prepend'] = buildImportStatement(enums, importFrom);
  }
  return result;
};

export default { plugin };
