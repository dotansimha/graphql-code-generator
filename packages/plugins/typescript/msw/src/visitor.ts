import {
  ClientSideBaseVisitor,
  ClientSideBasePluginConfig,
  getConfigValue,
  LoadedFragment,
} from '@graphql-codegen/visitor-plugin-common';
import { MSWRawPluginConfig } from './config';
import autoBind from 'auto-bind';
import { OperationDefinitionNode, GraphQLSchema, print } from 'graphql';
import { pascalCase } from 'change-case-all';

export interface MSWPluginConfig extends ClientSideBasePluginConfig {
  link?: {
    endpoint: string;
    name: string;
  };
}

export class MSWVisitor extends ClientSideBaseVisitor<MSWRawPluginConfig, MSWPluginConfig> {
  private _operationsToInclude: {
    node: OperationDefinitionNode;
    documentVariableName: string;
    operationType: string;
    operationResultType: string;
    operationVariablesTypes: string;
  }[] = [];

  constructor(schema: GraphQLSchema, fragments: LoadedFragment[], rawConfig: MSWRawPluginConfig) {
    super(schema, fragments, rawConfig, { link: getConfigValue(rawConfig.link, undefined) });

    autoBind(this);
  }

  public getImports(): string[] {
    const hasOperations = this._collectedOperations.length > 0;

    if (!hasOperations) {
      return [];
    }

    return [`import { graphql, ResponseResolver, GraphQLRequest, GraphQLContext } from 'msw'`];
  }

  public getContent() {
    const { link } = this.config;
    let endpoint: string;

    if (link) {
      endpoint = `const ${link.name} = graphql.link('${link.endpoint}')\n`;
    }

    const suffix = pascalCase(link?.name || '');
    const operations = this._operationsToInclude.map(
      ({ node, operationType, operationResultType, operationVariablesTypes }) => {
        if (operationType === 'Query' || operationType === 'Mutation') {
          const handlerName = `mock${pascalCase(node.name.value)}${operationType}${suffix}`;

          /** @ts-expect-error name DOES exist on @type{import('graphql').SelectionNode} */
          const selections = node.selectionSet.selections.map(sel => sel.name.value).join(', ');
          const variables = node.variableDefinitions.map(def => def.variable.name.value).join(', ');

          return `/**
 * @param resolver a function that accepts a captured request and may return a mocked response.
 * @see https://mswjs.io/docs/basics/response-resolver
 * @example
 * ${handlerName}((req, res, ctx) => {${variables && `\n *   const { ${variables} } = req.variables;`}
 *   return res(
 *     ctx.data({ ${selections} })
 *   )
 * })
 */
export const ${handlerName} = (resolver: ResponseResolver<GraphQLRequest<${operationVariablesTypes}>, GraphQLContext<${operationResultType}>, any>) =>
  ${link?.name || 'graphql'}.${operationType.toLowerCase()}<${operationResultType}, ${operationVariablesTypes}>(
    '${node.name.value}',
    resolver
  )\n`;
        }
        return '';
      }
    );

    return [endpoint, ...operations].join('\n');
  }

  buildOperation(node, documentVariableName, operationType, operationResultType, operationVariablesTypes) {
    if (node.name == null) {
      throw new Error("Plugin 'msw' cannot generate mocks for unnamed operation.\n\n" + print(node));
    } else {
      this._operationsToInclude.push({
        node,
        documentVariableName,
        operationType,
        operationResultType,
        operationVariablesTypes,
      });
    }

    return null;
  }
}
