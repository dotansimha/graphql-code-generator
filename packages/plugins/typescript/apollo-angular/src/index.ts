import { Types, PluginValidateFn, PluginFunction } from '@graphql-codegen/plugin-helpers';
import { visit, GraphQLSchema, concatAST, Kind, FragmentDefinitionNode, OperationDefinitionNode } from 'graphql';
import { RawClientSideBasePluginConfig, LoadedFragment } from '@graphql-codegen/visitor-plugin-common';
import { ApolloAngularVisitor } from './visitor';
import { extname } from 'path';
import gql from 'graphql-tag';

export interface ApolloAngularRawPluginConfig extends RawClientSideBasePluginConfig {
  /**
   * @name ngModule
   * @type string
   * @description Allows to define `ngModule` as part of the plugin's config so it's globally available.
   *
   * @example graphql.macro
   * ```yml
   * config:
   *   ngModule: ./path/to/module#MyModule
   * ```
   */
  ngModule?: string;
  /**
   * @name namedClient
   * @type string
   * @description Defined the global value of `namedClient`.
   *
   * @example graphql.macro
   * ```yml
   * config:
   *   namedClient: 'customName'
   * ```
   */
  namedClient?: string;
}

export const plugin: PluginFunction<ApolloAngularRawPluginConfig> = (schema: GraphQLSchema, documents: Types.DocumentFile[], config) => {
  const allAst = concatAST(
    documents.reduce((prev, v) => {
      return [...prev, v.content];
    }, [])
  );
  const operations = allAst.definitions.filter(d => d.kind === Kind.OPERATION_DEFINITION) as OperationDefinitionNode[];
  const allFragments: LoadedFragment[] = [
    ...(allAst.definitions.filter(d => d.kind === Kind.FRAGMENT_DEFINITION) as FragmentDefinitionNode[]).map(fragmentDef => ({ node: fragmentDef, name: fragmentDef.name.value, onType: fragmentDef.typeCondition.name.value, isExternal: false })),
    ...(config.externalFragments || []),
  ];

  const visitor = new ApolloAngularVisitor(schema, allFragments, operations, config) as any;
  const visitorResult = visit(allAst, { leave: visitor });

  return {
    prepend: visitor.getImports(),
    content: [visitor.fragments, ...visitorResult.definitions.filter(t => typeof t === 'string')].join('\n'),
  };
};

export const addToSchema = gql`
  directive @NgModule(module: String!) on OBJECT | FIELD
  directive @namedClient(name: String!) on OBJECT | FIELD
`;

export const validate: PluginValidateFn<any> = async (schema: GraphQLSchema, documents: Types.DocumentFile[], config, outputFile: string) => {
  if (extname(outputFile) !== '.ts') {
    throw new Error(`Plugin "apollo-angular" requires extension to be ".ts"!`);
  }
};

export { ApolloAngularVisitor };
