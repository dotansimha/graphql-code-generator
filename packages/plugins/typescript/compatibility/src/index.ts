import { GraphQLSchema, concatAST, visit } from 'graphql';
import { PluginFunction, Types } from '@graphql-codegen/plugin-helpers';
import { CompatibilityPluginVisitor } from './visitor';
import { CompatibilityPluginRawConfig } from './config';

const REACT_APOLLO_PLUGIN_NAME = 'typescript-react-apollo';

export const plugin: PluginFunction<CompatibilityPluginRawConfig> = async (
  schema: GraphQLSchema,
  documents: Types.DocumentFile[],
  config: CompatibilityPluginRawConfig,
  additionalData
): Promise<string> => {
  const allAst = concatAST(documents.map(v => v.document));

  const reactApollo = ((additionalData || {}).allPlugins || []).find(
    p => Object.keys(p)[0] === REACT_APOLLO_PLUGIN_NAME
  );
  const visitor = new CompatibilityPluginVisitor(config, schema, {
    reactApollo: reactApollo
      ? {
          ...(config || {}),
          ...(reactApollo[REACT_APOLLO_PLUGIN_NAME] as object),
        }
      : null,
  });

  const visitorResult = visit(allAst, {
    leave: visitor as any,
  });

  const discriminateUnion = `type DiscriminateUnion<T, U> = T extends U ? T : never;\n`;
  const result: string = visitorResult.definitions.filter(a => a && typeof a === 'string').join('\n');

  return result.includes('DiscriminateUnion') ? [discriminateUnion, result].join('\n') : result;
};
