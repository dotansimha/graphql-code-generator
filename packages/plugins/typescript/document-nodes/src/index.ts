import { PluginFunction, PluginValidateFn, Types } from '@graphql-codegen/plugin-helpers';
import {
  NamingConvention,
  LoadedFragment,
  RawClientSideBasePluginConfig,
} from '@graphql-codegen/visitor-plugin-common';
import { GraphQLSchema, visit, concatAST, FragmentDefinitionNode, Kind } from 'graphql';
import { TypeScriptDocumentNodesVisitor } from './visitor';

/**
 * @description This plugin generates TypeScript source (`.ts`) file from GraphQL files (`.graphql`).
 */
export interface TypeScriptDocumentNodesRawPluginConfig extends RawClientSideBasePluginConfig {
  /**
   * @default change-case-all#pascalCase
   * @description Allow you to override the naming convention of the output.
   * You can either override all namings, or specify an object with specific custom naming convention per output.
   * The format of the converter must be a valid `module#method`.
   * Allowed values for specific output are: `typeNames`, `enumValues`.
   * You can also use "keep" to keep all GraphQL names as-is.
   * Additionally you can set `transformUnderscore` to `true` if you want to override the default behavior,
   * which is to preserves underscores.
   *
   * Available case functions in `change-case-all` are `camelCase`, `capitalCase`, `constantCase`, `dotCase`, `headerCase`, `noCase`, `paramCase`, `pascalCase`, `pathCase`, `sentenceCase`, `snakeCase`, `lowerCase`, `localeLowerCase`, `lowerCaseFirst`, `spongeCase`, `titleCase`, `upperCase`, `localeUpperCase` and `upperCaseFirst`
   * [See more](https://github.com/btxtiger/change-case-all)
   *
   * @exampleMarkdown
   * ## Override All Names
   * ```yml
   * config:
   *   namingConvention: change-case-all#lowerCase
   * ```
   *
   * ## Upper-case enum values
   * ```yml
   * config:
   *   namingConvention:
   *     typeNames: change-case-all#pascalCase
   *     enumValues: change-case-all#upperCase
   * ```
   *
   * ## Keep name as-is
   * ```yml
   * config:
   *   namingConvention: keep
   * ```
   *
   * ## Remove Underscores
   * ```yml
   * config:
   *   namingConvention:
   *     typeNames: change-case-all#pascalCase
   *     transformUnderscore: true
   * ```
   */
  namingConvention?: NamingConvention;
  /**
   * @default ""
   * @description Adds prefix to the name
   *
   * @exampleMarkdown
   * ```yml
   *  documents: src/api/user-service/queries.graphql
   *  generates:
   *    src/api/user-service/queries.ts:
   *    plugins:
   *      - typescript-document-nodes
   *    config:
   *      namePrefix: 'gql'
   * ```
   */
  namePrefix?: string;
  /**
   * @default ""
   * @description Adds suffix to the name
   *
   * @exampleMarkdown
   * ```yml
   *  documents: src/api/user-service/queries.graphql
   *  generates:
   *    src/api/user-service/queries.ts:
   *    plugins:
   *      - typescript-document-nodes
   *    config:
   *      nameSuffix: 'Query'
   * ```
   */
  nameSuffix?: string;
  /**
   * @default ""
   * @description Adds prefix to the fragment variable
   */
  fragmentPrefix?: string;
  /**
   * @default ""
   * @description Adds suffix to the fragment variable
   */
  fragmentSuffix?: string;
}

export const plugin: PluginFunction<TypeScriptDocumentNodesRawPluginConfig> = (
  schema: GraphQLSchema,
  documents: Types.DocumentFile[],
  config: TypeScriptDocumentNodesRawPluginConfig
) => {
  const allAst = concatAST(documents.map(v => v.document));

  const allFragments: LoadedFragment[] = [
    ...(allAst.definitions.filter(d => d.kind === Kind.FRAGMENT_DEFINITION) as FragmentDefinitionNode[]).map(
      fragmentDef => ({
        node: fragmentDef,
        name: fragmentDef.name.value,
        onType: fragmentDef.typeCondition.name.value,
        isExternal: false,
      })
    ),
    ...(config.externalFragments || []),
  ];

  const visitor = new TypeScriptDocumentNodesVisitor(schema, allFragments, config, documents);
  const visitorResult = visit(allAst, { leave: visitor });

  return {
    prepend: visitor.getImports(),
    content: [visitor.fragments, ...visitorResult.definitions.filter(t => typeof t === 'string')].join('\n'),
  };
};

export const validate: PluginValidateFn<any> = async (
  schema: GraphQLSchema,
  documents: Types.DocumentFile[],
  config: any,
  outputFile: string
) => {
  if (!outputFile.endsWith('.ts')) {
    throw new Error(`Plugin "typescript-document-nodes" requires extension to be ".ts"!`);
  }
};
