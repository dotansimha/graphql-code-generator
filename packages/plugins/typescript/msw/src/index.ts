import { Types, PluginValidateFn, PluginFunction, oldVisit } from '@graphql-codegen/plugin-helpers';
import { GraphQLSchema, concatAST, Kind, FragmentDefinitionNode } from 'graphql';
import { LoadedFragment } from '@graphql-codegen/visitor-plugin-common';
import { MSWVisitor } from './visitor';
import { extname } from 'path';
import { MSWRawPluginConfig } from './config';

export const plugin: PluginFunction<MSWRawPluginConfig, Types.ComplexPluginOutput> = (schema, documents, config) => {
  const allAst = concatAST(documents.map(v => v.document));
  const allFragments: LoadedFragment[] = [
    ...(allAst.definitions.filter(d => d.kind === Kind.FRAGMENT_DEFINITION) as FragmentDefinitionNode[]).map(
      fragmentDef => ({
        node: fragmentDef,
        name: fragmentDef.name.value,
        onType: fragmentDef.typeCondition.name.value,
        isExternal: false,
      })
    ),
    ...(config.externalFragments || []),
  ];
  const visitor = new MSWVisitor(schema, allFragments, config);
  oldVisit(allAst, { leave: visitor });

  return {
    prepend: visitor.getImports(),
    content: visitor.getContent(),
  };
};

export const validate: PluginValidateFn<any> = async (
  schema: GraphQLSchema,
  documents: Types.DocumentFile[],
  config: MSWRawPluginConfig,
  outputFile: string
) => {
  if (extname(outputFile) !== '.ts') {
    throw new Error(`Plugin "typescript-msw" requires extension to be ".ts"!`);
  }
};

export { MSWVisitor };
