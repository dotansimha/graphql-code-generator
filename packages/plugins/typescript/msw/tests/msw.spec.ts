import { pascalCase } from 'change-case-all';
import { parse } from 'graphql';
import { plugin } from '../src';

describe('msw', () => {
  const queryName = 'User';
  const mutationName = 'UpdateUser';
  const documents = [
    { document: parse(`query ${queryName} { name }`) },
    { document: parse(`mutation ${mutationName} { name }`) },
  ];

  it('Should generate mocks based on queries and mutations', async () => {
    const result = await plugin(null, documents, {});

    // handler function names
    expect(result.content).toContain(`mock${queryName}Query`);
    expect(result.content).toContain(`mock${mutationName}Mutation`);

    // handler strings
    expect(result.content).toContain(`'${queryName}',`);
    expect(result.content).toContain(`'${mutationName}',`);

    expect(result.prepend).toMatchSnapshot('imports');
    expect(result.content).toMatchSnapshot('content');
  });

  it('Should generate a link with an endpoint when passed a link variable', async () => {
    const link = { name: 'api', endpoint: 'http://localhost:3000/graphql' };
    const capitalizedName = pascalCase(link.name);

    const result = await plugin(null, documents, { link });

    // endpoint definition
    expect(result.content).toContain(`const ${link.name} = graphql.link('${link.endpoint}')`);

    // endpoint used instead of `graphql`
    expect(result.content).not.toContain(`graphql.query`);
    expect(result.content).not.toContain(`graphql.mutation`);
    expect(result.content).toContain(`${link.name}.query`);
    expect(result.content).toContain(`${link.name}.mutation`);

    // handler function names
    expect(result.content).toContain(`mock${queryName}Query${capitalizedName}`);
    expect(result.content).toContain(`mock${mutationName}Mutation${capitalizedName}`);

    expect(result.content).toMatchSnapshot('content with a link/endpoint');
  });

  it('Should generate JSDoc documentation with variables and selection from the operations themselves', async () => {
    const variables = ['id', 'offset', 'limit'];
    const input = variables.map(v => `$${v}: Int`).join(', ');

    const fields = ['name', 'phone', 'country'];
    const selection = fields.join(', ');

    const documents = [
      { document: parse(`query ${queryName}(${input}) { ${selection} }`) },
      { document: parse(`mutation ${mutationName}(${input}) { ${selection} }`) },
    ];

    const result = await plugin(null, documents, {});

    expect(result.content).toContain(`const { ${variables.join(', ')} } = req.variables`);
    expect(result.content).toContain(`ctx.data({ ${selection} })`);
    expect(result.content).toMatchSnapshot('content with variables and selection JSDoc documentation');
  });
});
