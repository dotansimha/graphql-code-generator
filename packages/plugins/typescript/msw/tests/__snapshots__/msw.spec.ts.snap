// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`msw Should generate JSDoc documentation with variables and selection from the operations themselves: content with variables and selection JSDoc documentation 1`] = `
"
/**
 * @param resolver a function that accepts a captured request and may return a mocked response.
 * @see https://mswjs.io/docs/basics/response-resolver
 * @example
 * mockUserQuery((req, res, ctx) => {
 *   const { id, offset, limit } = req.variables;
 *   return res(
 *     ctx.data({ name, phone, country })
 *   )
 * })
 */
export const mockUserQuery = (resolver: ResponseResolver<GraphQLRequest<UserQueryVariables>, GraphQLContext<UserQuery>, any>) =>
  graphql.query<UserQuery, UserQueryVariables>(
    'User',
    resolver
  )

/**
 * @param resolver a function that accepts a captured request and may return a mocked response.
 * @see https://mswjs.io/docs/basics/response-resolver
 * @example
 * mockUpdateUserMutation((req, res, ctx) => {
 *   const { id, offset, limit } = req.variables;
 *   return res(
 *     ctx.data({ name, phone, country })
 *   )
 * })
 */
export const mockUpdateUserMutation = (resolver: ResponseResolver<GraphQLRequest<UpdateUserMutationVariables>, GraphQLContext<UpdateUserMutation>, any>) =>
  graphql.mutation<UpdateUserMutation, UpdateUserMutationVariables>(
    'UpdateUser',
    resolver
  )
"
`;

exports[`msw Should generate a link with an endpoint when passed a link variable: content with a link/endpoint 1`] = `
"const api = graphql.link('http://localhost:3000/graphql')

/**
 * @param resolver a function that accepts a captured request and may return a mocked response.
 * @see https://mswjs.io/docs/basics/response-resolver
 * @example
 * mockUserQueryApi((req, res, ctx) => {
 *   return res(
 *     ctx.data({ name })
 *   )
 * })
 */
export const mockUserQueryApi = (resolver: ResponseResolver<GraphQLRequest<UserQueryVariables>, GraphQLContext<UserQuery>, any>) =>
  api.query<UserQuery, UserQueryVariables>(
    'User',
    resolver
  )

/**
 * @param resolver a function that accepts a captured request and may return a mocked response.
 * @see https://mswjs.io/docs/basics/response-resolver
 * @example
 * mockUpdateUserMutationApi((req, res, ctx) => {
 *   return res(
 *     ctx.data({ name })
 *   )
 * })
 */
export const mockUpdateUserMutationApi = (resolver: ResponseResolver<GraphQLRequest<UpdateUserMutationVariables>, GraphQLContext<UpdateUserMutation>, any>) =>
  api.mutation<UpdateUserMutation, UpdateUserMutationVariables>(
    'UpdateUser',
    resolver
  )
"
`;

exports[`msw Should generate mocks based on queries and mutations: content 1`] = `
"
/**
 * @param resolver a function that accepts a captured request and may return a mocked response.
 * @see https://mswjs.io/docs/basics/response-resolver
 * @example
 * mockUserQuery((req, res, ctx) => {
 *   return res(
 *     ctx.data({ name })
 *   )
 * })
 */
export const mockUserQuery = (resolver: ResponseResolver<GraphQLRequest<UserQueryVariables>, GraphQLContext<UserQuery>, any>) =>
  graphql.query<UserQuery, UserQueryVariables>(
    'User',
    resolver
  )

/**
 * @param resolver a function that accepts a captured request and may return a mocked response.
 * @see https://mswjs.io/docs/basics/response-resolver
 * @example
 * mockUpdateUserMutation((req, res, ctx) => {
 *   return res(
 *     ctx.data({ name })
 *   )
 * })
 */
export const mockUpdateUserMutation = (resolver: ResponseResolver<GraphQLRequest<UpdateUserMutationVariables>, GraphQLContext<UpdateUserMutation>, any>) =>
  graphql.mutation<UpdateUserMutation, UpdateUserMutationVariables>(
    'UpdateUser',
    resolver
  )
"
`;

exports[`msw Should generate mocks based on queries and mutations: imports 1`] = `
Array [
  "import { graphql, ResponseResolver, GraphQLRequest, GraphQLContext } from 'msw'",
]
`;
